# Cohort_ver22_complete_STEMI
## 진행 사항
- cohort_ver21의 컬럼명 변경
- datatime으로 통일
- 결측치 확인 및 보완





구글 드라이브 마운트

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# (1) cohort_ver21의 컬럼명 변경

In [ ]:
import pandas as pd
import os

# 본인 작업 환경에 맞게 경로 수정 후 실행
base_path='/content/drive/MyDrive/Colab Notebooks/2025-2 데이터애널리틱스'

cohort_ver21 = pd.read_csv(os.path.join(base_path, 'cohort_ver21_expanded_STEMI(2).csv'))
# cohort_ver21의 컬럼명 변경
cohort_ver21.rename(columns={'stay_id':'icu_stay_id','intime':'icu_intime', 'outtime':'icu_outtime', 'los_days':'icu_los', 'first_ecg_time':'first_stemi_ecg_time'}, inplace=True)
print(cohort_ver21.columns)
cohort_ver21.head()

Index(['subject_id', 'hadm_id', 'icu_stay_id', 'icu_intime', 'icu_outtime',
       'gender', 'race', 'arrival_transport', 'disposition',
       'first_stemi_ecg_time', 'door_to_ecg_min', 'admittime', 'lab_tat_min',
       'door_to_antithrombotic_min', 'door_to_cath_min', 'boarding_delay_min',
       'death_flag', 'icu_los', 'transfer_type', 'is_emergency_admission',
       'dod', 'eventtype', 'cohort_source'],
      dtype='object')


subject_id     hadm_id  icu_stay_id                 icu_intime  \
0    10010058  26359957.0     39919549        2147-11-18 00:50:00   
1    10012438  22764825.0     35750515        2178-06-07 19:33:00   
2    10013310  21243435.0     31736720        2153-05-26 08:56:00   
3    10013310  27682188.0     35160955  2153-05-06 10:21:00+00:00   
4    10033552  26487381.0     32542054  2136-02-29 22:09:00+00:00   

                 icu_outtime gender                      race  \
0        2147-11-18 03:19:00      M  ASIAN - SOUTH EAST ASIAN   
1        2178-06-07 21:57:00      M                     WHITE   
2        2153-05-26 14:18:39      F             BLACK/AFRICAN   
3  2153-05-06 18:28:00+00:00      F             BLACK/AFRICAN   
4  2136-03-01 02:26:37+00:00      F                     OTHER   

  arrival_transport disposition first_stemi_ecg_time  door_to_ecg_min  \
0         AMBULANCE    ADMITTED                  NaN              NaN   
1         AMBULANCE    ADMITTED                  NaN              NaN   
2           UNKNOWN    ADMITTED                  NaN              NaN   
3         AMBULANCE    ADMITTED     2153-05-06 10:25              4.0   
4         AMBULANCE    ADMITTED     2136-02-29 22:21             12.0   

                   admittime  lab_tat_min  door_to_antithrombotic_min  \
0        2147-11-18 02:34:00          NaN                         NaN   
1        2179-07-12 05:56:00          NaN                         NaN   
2        2153-06-10 11:55:00          NaN                         NaN   
3  2153-05-06 18:03:00+00:00         61.0                         NaN   
4  2136-03-01 02:25:00+00:00        107.0                         NaN   

   door_to_cath_min  boarding_delay_min  death_flag   icu_los  \
0               NaN                 NaN         1.0  1.232326   
1               NaN                 NaN         0.0  0.746007   
2               NaN                 NaN         1.0  6.296898   
3             484.0          -25.000000         1.0  6.820000   
4               NaN           -1.616667         1.0  2.750000   

      transfer_type  is_emergency_admission         dod  eventtype  \
0               NaN                     NaN  2147-11-19  discharge   
1               NaN                     NaN         NaN  discharge   
2               NaN                     NaN  2153-11-19  discharge   
3  HOME HEALTH CARE                     1.0  2153-11-19  discharge   
4              HOME                     0.0  2141-07-15  discharge   

   cohort_source  
0  ICD_Confirmed  
1  ICD_Confirmed  
2  ICD_Confirmed  
3  ICD_Confirmed  
4  ICD_Confirmed

# (2) datatime으로 통일

In [ ]:
for col in ['icu_intime', 'icu_outtime', 'admittime']:
  cohort_ver21[col] = pd.to_datetime(cohort_ver21[col], errors='coerce')

pd.set_option('display.max_columns', None)
cohort_ver21.head()

subject_id     hadm_id  icu_stay_id          icu_intime  \
0    10010058  26359957.0     39919549 2147-11-18 00:50:00   
1    10012438  22764825.0     35750515 2178-06-07 19:33:00   
2    10013310  21243435.0     31736720 2153-05-26 08:56:00   
3    10013310  27682188.0     35160955                 NaT   
4    10033552  26487381.0     32542054                 NaT   

          icu_outtime gender                      race arrival_transport  \
0 2147-11-18 03:19:00      M  ASIAN - SOUTH EAST ASIAN         AMBULANCE   
1 2178-06-07 21:57:00      M                     WHITE         AMBULANCE   
2 2153-05-26 14:18:39      F             BLACK/AFRICAN           UNKNOWN   
3                 NaT      F             BLACK/AFRICAN         AMBULANCE   
4                 NaT      F                     OTHER         AMBULANCE   

  disposition first_stemi_ecg_time  door_to_ecg_min           admittime  \
0    ADMITTED                  NaN              NaN 2147-11-18 02:34:00   
1    ADMITTED                  NaN              NaN 2179-07-12 05:56:00   
2    ADMITTED                  NaN              NaN 2153-06-10 11:55:00   
3    ADMITTED     2153-05-06 10:25              4.0                 NaT   
4    ADMITTED     2136-02-29 22:21             12.0                 NaT   

   lab_tat_min  door_to_antithrombotic_min  door_to_cath_min  \
0          NaN                         NaN               NaN   
1          NaN                         NaN               NaN   
2          NaN                         NaN               NaN   
3         61.0                         NaN             484.0   
4        107.0                         NaN               NaN   

   boarding_delay_min  death_flag   icu_los     transfer_type  \
0                 NaN         1.0  1.232326               NaN   
1                 NaN         0.0  0.746007               NaN   
2                 NaN         1.0  6.296898               NaN   
3          -25.000000         1.0  6.820000  HOME HEALTH CARE   
4           -1.616667         1.0  2.750000              HOME   

   is_emergency_admission         dod  eventtype  cohort_source  
0                     NaN  2147-11-19  discharge  ICD_Confirmed  
1                     NaN         NaN  discharge  ICD_Confirmed  
2                     NaN  2153-11-19  discharge  ICD_Confirmed  
3                     1.0  2153-11-19  discharge  ICD_Confirmed  
4                     0.0  2141-07-15  discharge  ICD_Confirmed

# (3) 결측치 확인 및 일부 보완

- 결측치 확인

In [ ]:
print(cohort_ver21.isnull().sum())

subject_id                       0
hadm_id                         53
icu_stay_id                      0
icu_intime                     465
icu_outtime                    465
gender                           0
race                            53
arrival_transport             1147
disposition                      0
first_stemi_ecg_time          1518
door_to_ecg_min               1518
admittime                     1612
lab_tat_min                   1549
door_to_antithrombotic_min    1681
door_to_cath_min              1706
boarding_delay_min            1518
death_flag                    1147
icu_los                       1221
transfer_type                 1518
is_emergency_admission        1518
dod                           1686
eventtype                     1147
cohort_source                    0
dtype: int64


## 3.1. hadm_id 결측치 확인 및 보완
- STEMI 환자의 경우 입원을 하거나 / ed에서 사망하거나 둘 중 하나
- 따라서, hadm_id가 없는 환자는 stemi가 아니거나 / ed에서 사망했을 확률이 높으므로 hadm_id가 없는 row만 분석

### 3.1.1 입원 정보가 없는 환자 수 파악

In [ ]:
no_hadm_id = cohort_ver21[cohort_ver21['hadm_id'].isnull()]
print(f'입원 정보가 없는 환자 수: {len(no_hadm_id)}명')
no_hadm_id

입원 정보가 없는 환자 수: 53명


subject_id  hadm_id  icu_stay_id          icu_intime  \
918     10692509      NaN     31516860 2170-02-13 10:04:00   
933     10803632      NaN     37430690 2200-03-26 22:17:00   
948     10944352      NaN     34286973 2125-12-26 16:01:00   
949     10947511      NaN     30740461 2168-06-19 04:46:00   
959     11020337      NaN     33735546 2198-04-24 14:38:00   
963     11037551      NaN     31114742 2193-07-28 12:22:00   
982     11238662      NaN     31106614 2136-01-21 14:37:00   
985     11258297      NaN     32822230 2153-12-05 06:59:00   
1008    11478064      NaN     34404806 2133-01-27 09:26:00   
1024    11619103      NaN     38102175 2118-07-03 23:39:00   
1026    11645882      NaN     32226427 2179-04-02 08:54:00   
1045    11875213      NaN     36563288 2184-08-13 21:26:00   
1082    12247387      NaN     32563598 2163-03-31 22:02:00   
1091    12306602      NaN     30167252 2131-04-23 12:53:00   
1109    12497922      NaN     30410080 2130-12-08 10:01:00   
1126    12700442      NaN     35367265 2149-05-13 09:55:00   
1132    12739018      NaN     38841772 2158-10-06 11:05:00   
1151    12887630      NaN     38989120 2142-08-11 02:31:00   
1227    13496430      NaN     39878729 2144-03-05 15:30:00   
1234    13536343      NaN     39730329 2127-05-28 01:32:00   
1292    14028344      NaN     34325425 2168-07-25 05:44:00   
1318    14264560      NaN     35582912 2163-12-04 01:10:00   
1347    14530934      NaN     33360072 2139-05-26 08:41:00   
1367    14721603      NaN     38105398 2178-07-12 10:06:00   
1405    15101183      NaN     37993213 2194-11-19 04:02:00   
1435    15259624      NaN     30458358 2130-07-17 00:10:00   
1437    15293902      NaN     32979309 2118-10-24 23:05:00   
1441    15317910      NaN     39518063 2147-08-12 15:56:00   
1455    15451693      NaN     32774155 2136-08-08 19:02:00   
1501    15796783      NaN     33400722 2111-01-25 02:43:00   
1506    15826743      NaN     34588202 2150-01-21 09:46:00   
1509    15834787      NaN     30505441 2152-09-06 17:40:00   
1520    15929083      NaN     35229277 2182-09-11 18:32:00   
1524    15976052      NaN     39751346 2170-06-05 18:49:00   
1598    16610408      NaN     38105323 2189-08-24 15:56:00   
1614    16816345      NaN     36747034 2179-01-02 22:13:00   
1639    16976120      NaN     35080832 2191-12-28 03:46:00   
1654    17114933      NaN     38413885 2151-05-24 22:46:00   
1721    17683789      NaN     39637534 2180-02-23 15:29:00   
1738    17867860      NaN     33183879 2132-04-02 16:11:00   
1770    18164954      NaN     39223977 2144-11-05 09:55:00   
1777    18226953      NaN     33567746 2146-05-22 13:04:00   
1837    18736648      NaN     39600698 2176-06-20 20:54:00   
1879    19190224      NaN     32806578 2188-03-14 15:51:00   
1885    19225934      NaN     37373089 2160-10-16 13:37:00   
1893    19281541      NaN     39211923 2121-09-26 17:09:00   
1900    19355082      NaN     32116089 2152-09-20 11:27:00   
1927    19618591      NaN     34817477 2205-05-07 08:51:00   
1940    19662153      NaN     39952995 2147-02-21 13:43:00   
1959    19792891      NaN     36898083 2170-10-01 02:24:00   
1973    19922851      NaN     37942091 2197-02-02 10:47:00   
1975    19925436      NaN     37513555 2129-07-12 19:59:00   
1982    19993842      NaN     37776073 2148-12-08 17:23:00   

             icu_outtime gender race arrival_transport  \
918  2170-02-13 15:08:00      M  NaN               NaN   
933  2200-03-27 04:54:00      M  NaN               NaN   
948  2125-12-26 17:41:00      F  NaN               NaN   
949  2168-06-19 09:21:00      M  NaN               NaN   
959  2198-04-24 22:57:00      M  NaN               NaN   
963  2193-07-28 17:09:00      M  NaN               NaN   
982  2136-01-21 15:58:00      M  NaN               NaN   
985  2153-12-05 14:02:00      F  NaN               NaN   
1008 2133-01-27 15:12:00      M  NaN               NaN   
1024 2118-07-04 01:44:00      F  NaN               NaN   
1026 2179-04-02 18:

### 3.1.2 입원 정보가 없는 환자의 응급실 사망 가능성 확인
- 'dispotition' 분포 및 'death_flag' 확인

In [ ]:
print("------ 'disposition'분포 ------")
print(no_hadm_id['disposition'].value_counts())

print("------ 'death_flag'분포 ------")
print(no_hadm_id['death_flag'].value_counts(dropna=False))

------ 'disposition'분포 ------
disposition
EXPIRED                        12
TRANSFER                       12
ELOPED                         11
LEFT WITHOUT BEING SEEN         9
LEFT AGAINST MEDICAL ADVICE     7
ADMITTED                        1
OTHER                           1
Name: count, dtype: int64
------ 'death_flag'분포 ------
death_flag
NaN    53
Name: count, dtype: int64


- death_flag는 53명 모두 NaN

- disposition 분포 결과 분석  

1. 응급실에서 사망(EXPIRED): 12명  
2. 전원(TRANSFER): 12명  
3. 동의 없이 병원 떠남(ELOPED): 11명  
4. 진료를 받지 않고 떠남(LEFT WITHOUT BEING SEEN): 9명  
5. 의학적 권고에 반하여 퇴원(LEFT AGAINST MEDICAL ADVICE): 7명  
6. 응급실에 입원했으나 hadm_id가 누락된 데이터 불일치 코드(ADMITTED): 1명  

-> 1~5번의 환자들은 stemi라 하더라도 처치 기록이 없으므로 cohort에서 제거  
-> 6번 환자 역시 hadm_id 결측은 이후 분석에 지장이 가므로 제거

### 3.1.3 입원 정보가 없는 환자들을 코호트에서 제외  
1983 -> 1930명

In [ ]:
cohort_ver21_hadm = cohort_ver21.dropna(subset=['hadm_id'])
cohort_ver21_hadm.head()

subject_id     hadm_id  icu_stay_id          icu_intime  \
0    10010058  26359957.0     39919549 2147-11-18 00:50:00   
1    10012438  22764825.0     35750515 2178-06-07 19:33:00   
2    10013310  21243435.0     31736720 2153-05-26 08:56:00   
3    10013310  27682188.0     35160955                 NaT   
4    10033552  26487381.0     32542054                 NaT   

          icu_outtime gender                      race arrival_transport  \
0 2147-11-18 03:19:00      M  ASIAN - SOUTH EAST ASIAN         AMBULANCE   
1 2178-06-07 21:57:00      M                     WHITE         AMBULANCE   
2 2153-05-26 14:18:39      F             BLACK/AFRICAN           UNKNOWN   
3                 NaT      F             BLACK/AFRICAN         AMBULANCE   
4                 NaT      F                     OTHER         AMBULANCE   

  disposition first_stemi_ecg_time  door_to_ecg_min           admittime  \
0    ADMITTED                  NaN              NaN 2147-11-18 02:34:00   
1    ADMITTED                  NaN              NaN 2179-07-12 05:56:00   
2    ADMITTED                  NaN              NaN 2153-06-10 11:55:00   
3    ADMITTED     2153-05-06 10:25              4.0                 NaT   
4    ADMITTED     2136-02-29 22:21             12.0                 NaT   

   lab_tat_min  door_to_antithrombotic_min  door_to_cath_min  \
0          NaN                         NaN               NaN   
1          NaN                         NaN               NaN   
2          NaN                         NaN               NaN   
3         61.0                         NaN             484.0   
4        107.0                         NaN               NaN   

   boarding_delay_min  death_flag   icu_los     transfer_type  \
0                 NaN         1.0  1.232326               NaN   
1                 NaN         0.0  0.746007               NaN   
2                 NaN         1.0  6.296898               NaN   
3          -25.000000         1.0  6.820000  HOME HEALTH CARE   
4           -1.616667         1.0  2.750000              HOME   

   is_emergency_admission         dod  eventtype  cohort_source  
0                     NaN  2147-11-19  discharge  ICD_Confirmed  
1                     NaN         NaN  discharge  ICD_Confirmed  
2                     NaN  2153-11-19  discharge  ICD_Confirmed  
3                     1.0  2153-11-19  discharge  ICD_Confirmed  
4                     0.0  2141-07-15  discharge  ICD_Confirmed

## 3.2 icu_intime / icu_outtime 결측치 보완

In [ ]:
print(cohort_ver21_hadm.isnull().sum())

subject_id                       0
hadm_id                          0
icu_stay_id                      0
icu_intime                     465
icu_outtime                    465
gender                           0
race                             0
arrival_transport             1094
disposition                      0
first_stemi_ecg_time          1465
door_to_ecg_min               1465
admittime                     1559
lab_tat_min                   1496
door_to_antithrombotic_min    1628
door_to_cath_min              1653
boarding_delay_min            1465
death_flag                    1094
icu_los                       1168
transfer_type                 1465
is_emergency_admission        1465
dod                           1633
eventtype                     1094
cohort_source                    0
dtype: int64


### 3.2.1 결측 데이터 누락 경로 확인
- icu_intime, icu_outtime 결측치 465건이 데이터 병합 과정에서의 손실인지, 원본 데이터 누락인지 확인

In [ ]:
import pandas as pd
import os
import sqlite3
import numpy as np

base_path = '/content/drive/MyDrive/Colab Notebooks/2025-2 데이터애널리틱스'

# [1] 경로 설정

# ED 데이터셋 폴더 경로
ed_base_path = '/content/drive/MyDrive/Colab Notebooks/2025-2 데이터애널리틱스/mimic-iv-ed-2.2/'

# HOSP/ICU 데이터베이스 파일 경로
db_path = '/content/drive/MyDrive/Colab Notebooks/2025-2 데이터애널리틱스/MIMIC4-hosp-icu.db'

# [2] 데이터 로드 및 코호트 생성

# 2-1. ED 데이터 로드 (CSV)
edstays_csv_path = os.path.join(ed_base_path, 'edstays.csv')
df_edstays = pd.read_csv(edstays_csv_path,
                          dtype={'subject_id': int, 'hadm_id': 'Int64', 'stay_id': int})
print(f"edstays.csv 로드 완료 (Shape: {df_edstays.shape})")

# 2-2. HOSP 데이터 로드 (SQLite)
conn = sqlite3.connect(db_path) # DB 연결
query = "SELECT subject_id, hadm_id, icd_code, icd_version FROM diagnoses_icd"
df_diagnoses_icd = pd.read_sql_query(query, conn)
conn.close() # DB 연결 해제

df_diagnoses_icd['icd_code'] = df_diagnoses_icd['icd_code'].astype(str)
print(f"diagnoses_icd DB 테이블 로드 완료 (Shape: {df_diagnoses_icd.shape})")

# [2] ICU 데이터 원본 (icustays) 확인

try:
    conn = sqlite3.connect(db_path) # DB 연결

    # icustays 테이블은 ICU 입/퇴실 시간의 원본 데이터
    # subject_id, hadm_id, icu_intime, icu_outtime을 가져옴
    query_icu = """
    SELECT
        subject_id, hadm_id, stay_id as icu_stay_id, intime as icu_intime_db, outtime as icu_outtime_db
    FROM icustays
    """
    df_icustays = pd.read_sql_query(query_icu, conn)
    conn.close() # DB 연결 해제

    print(f"icustays DB 테이블 로드 완료 (Shape: {df_icustays.shape})")

    # 3. 현재 코호트와 icustays 테이블 병합 시도 (hadm_id 기준)
    # 현재 코호트에 누락된 ICU 기록이 있는지 확인하기 위함
    # hadm_id는 실수형이므로 형변환 (Int64)
    cohort_ver21_hadm['hadm_id'] = cohort_ver21_hadm['hadm_id'].astype('Int64')

    # 1:N 관계이므로, 하나의 hadm_id에 여러 icustay가 있을 수 있음
    df_check = pd.merge(
        cohort_ver21_hadm[['hadm_id', 'icu_intime']].dropna(subset=['hadm_id']),
        df_icustays[['hadm_id', 'icu_stay_id']].drop_duplicates(),
        on='hadm_id',
        how='left',
        indicator=True # 병합 결과 표시
    )

    # 4. ICU 기록이 누락된 행 확인
    # 현재 코호트에는 hadm_id가 있지만, icustays에는 없는 경우를 찾기
    # 'both'가 아닌 'left_only'이면서 icu_intime이 결측인 행의 수를 확인해야 함

    # 현재 코호트에서 icu_intime이 NaN인 행 (465개)을 추출
    df_missing_icu = cohort_ver21_hadm[cohort_ver21_hadm['icu_intime'].isnull()].copy()

    # icu_intime이 NaN인 행들의 hadm_id를 icustays 테이블에서 찾기
    missing_hadm_ids = df_missing_icu['hadm_id'].unique()

    # 이 hadm_id들이 icustays에 존재하는지 확인
    hadm_id_found_in_icu = df_icustays[df_icustays['hadm_id'].isin(missing_hadm_ids)]['hadm_id'].unique()


    print("\n--- ICU 결측치 병합 오류 검증 결과 ---")
    print(f"ICU 기록이 결측인 hadm_id 수 (현재 코호트 기준): {len(missing_hadm_ids)}개")
    print(f"ICU 기록이 결측이었지만 icustays 원본 테이블에서 발견된 hadm_id 수: {len(hadm_id_found_in_icu)}개")

    if len(hadm_id_found_in_icu) == 0:
        print("\n✅ 확인: ICU 기록이 결측인 모든 환자는 원본 icustays 테이블에도 기록이 없습니다.")
        print("결측치 465건은 병합 오류가 아닌 'ICU 미입원'을 의미할 가능성이 매우 높습니다.")
    else:
        print(f"\n⚠️ 경고: {len(hadm_id_found_in_icu)}개의 hadm_id가 ICU 결측 행이었으나, 원본 icustays 테이블에서 발견되었습니다.")
        print("이는 코호트 생성 또는 ICU 데이터 병합 과정에서 문제가 있었음을 의미합니다.")

# [5] 오류 처리 (DB 관련)
except pd.io.sql.DatabaseError as e:
    print(f"\n--- DB 연결 오류 (SQLite) ---")
    print(f"오류: DB 연결이나 'icustays' 테이블 쿼리에 실패했습니다.")
    print(f"경로: '{db_path}'")
    print(f"상세 오류: {e}")

except Exception as e:
    print(f"--- 알 수 없는 오류 ---")
    print(f"오류 발생: {e}")

edstays.csv 로드 완료 (Shape: (425087, 9))
diagnoses_icd DB 테이블 로드 완료 (Shape: (6364488, 4))
icustays DB 테이블 로드 완료 (Shape: (94458, 5))

--- ICU 결측치 병합 오류 검증 결과 ---
ICU 기록이 결측인 hadm_id 수 (현재 코호트 기준): 464개
ICU 기록이 결측이었지만 icustays 원본 테이블에서 발견된 hadm_id 수: 313개

⚠️ 경고: 313개의 hadm_id가 ICU 결측 행이었으나, 원본 icustays 테이블에서 발견되었습니다.
이는 코호트 생성 또는 ICU 데이터 병합 과정에서 문제가 있었음을 의미합니다.


/tmp/ipython-input-1942898676.py:53: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cohort_ver21_hadm['hadm_id'] = cohort_ver21_hadm['hadm_id'].astype('Int64')


### 3.2.2 ICU 입/퇴원 시간 결측치 보완

In [ ]:
import pandas as pd
import numpy as np

# 1. 병합을 위해 hadm_id 타입 통일 (Int64)
cohort_ver21_hadm['hadm_id'] = cohort_ver21_hadm['hadm_id'].astype('Int64')
df_icustays['hadm_id'] = df_icustays['hadm_id'].astype('Int64')

# 2. 중복 방지를 위해 환자별 '첫 번째' ICU 입원 기록만 추출
# 한 번의 입원(hadm_id) 중에 ICU를 여러 번 갔다 왔을 수 있으므로, 가장 먼저 들어간 시간(min)을 가져옴
icustays_first = df_icustays.sort_values('icu_intime_db').groupby('hadm_id').first().reset_index()

# 3. 코호트 데이터와 원본 ICU 데이터 병합 (Left Join)
# suffixes를 사용하여 기존 결측 컬럼과 구분
df_merged = pd.merge(
    cohort_ver21_hadm,
    icustays_first[['hadm_id', 'icu_intime_db', 'icu_outtime_db']], # 필요한 컬럼만
    on='hadm_id',
    how='left'
)

# 4. 결측치 채우기 (Data Imputation)
# 기존 icu_intime이 비어있다면(_original이 NaN), DB에서 가져온 값(_db)으로 채움
# 만약 기존 값도 있고 DB 값도 있다면, 기존 값을 우선합니다 (fillna는 결측일 때만 채움)

# 컬럼이 datetime 형식인지 확실히 하기 위해 변환
for col in ['icu_intime', 'icu_outtime']:
    df_merged[col] = pd.to_datetime(df_merged[col], errors='coerce')

df_merged['icu_intime_db'] = pd.to_datetime(df_merged['icu_intime_db'], errors='coerce')
df_merged['icu_outtime_db'] = pd.to_datetime(df_merged['icu_outtime_db'], errors='coerce')

# 결측치 채우기 (fillna)
df_merged['icu_intime'] = df_merged['icu_intime'].fillna(df_merged['icu_intime_db'])
df_merged['icu_outtime'] = df_merged['icu_outtime'].fillna(df_merged['icu_outtime_db'])

# 임시로 가져온 _db 컬럼 제거
df_merged = df_merged.drop(columns=['icu_intime_db', 'icu_outtime_db'])

# 5. 파생 변수 재계산 (is_icu_stay, icu_los_days)
# 이제 313명의 데이터가 채워졌으므로 플래그와 재원일수를 다시 계산
# ICU 입원 여부 (값이 있으면 1, 여전히 없으면 0)
df_merged['is_icu_stay'] = df_merged['icu_intime'].notna().astype(int)

# ICU 재원 일수 (분 단위 -> 일 단위)
df_merged['icu_los_days'] = (df_merged['icu_outtime'] - df_merged['icu_intime']).dt.total_seconds() / (60 * 60 * 24)

# 6. 결과 확인
print("--- ICU 데이터 복구 및 재계산 완료 ---")
print(f"총 행 수: {len(df_merged)}")

# 복구 후 결측치 확인
remaining_missing = df_merged['icu_intime'].isnull().sum()
print(f"복구 후 남은 ICU 결측치 수 (진짜 미입원): {remaining_missing}개")
print(f"복구된 데이터 수: {464 - remaining_missing}개 (예상치: 313개)")

# ICU 입원 여부 분포
print("\n최종 ICU 입원 여부(is_icu_stay) 분포:")
print(df_merged['is_icu_stay'].value_counts())

# 7. 변수명 업데이트 및 저장
cohort_ver22 = df_merged.copy()
print(cohort_ver22.columns)

/tmp/ipython-input-759648245.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cohort_ver21_hadm['hadm_id'] = cohort_ver21_hadm['hadm_id'].astype('Int64')


--- ICU 데이터 복구 및 재계산 완료 ---
총 행 수: 1930
복구 후 남은 ICU 결측치 수 (진짜 미입원): 151개
복구된 데이터 수: 313개 (예상치: 313개)

최종 ICU 입원 여부(is_icu_stay) 분포:
is_icu_stay
1    1779
0     151
Name: count, dtype: int64
Index(['subject_id', 'hadm_id', 'icu_stay_id', 'icu_intime', 'icu_outtime',
       'gender', 'race', 'arrival_transport', 'disposition',
       'first_stemi_ecg_time', 'door_to_ecg_min', 'admittime', 'lab_tat_min',
       'door_to_antithrombotic_min', 'door_to_cath_min', 'boarding_delay_min',
       'death_flag', 'icu_los', 'transfer_type', 'is_emergency_admission',
       'dod', 'eventtype', 'cohort_source', 'is_icu_stay', 'icu_los_days'],
      dtype='object')


### 결과 분석  
1. is_icu_stay = 0 인 151명은 hadm_id가 있어서 입원 기록은 존재하지만 icustays.csv 테이블에는 기록이 없음  
2. 이는 응급실을 거쳐 중환자실(ICU)이 아닌, 일반 병동 또는 심장내과 전문 병동에 입원한 환자들을 의미함  
3. STEMI 환자가 ICU에 가지 않는 경우:   
- 시술 후 상태가 안정적이어서 중환자실 감시가 필요 없다고 판단됨  
- ICU 병상 부족으로 준중환자실에서 치료  
- 시술 중 사망하거나 수술실로 직행하여 거기에서 사망한 경우

## 3.2.3 is_icu_stay = 0 인 151명의 death_flag 분석

### (1) 재원일수 확인 및 보완  
1. icu_los_days = 0 으로 채우기 (icu에 재원하지 않았으므로)  
2. los_hospital_days 재계산
3. 중증도를 파악하기 위해 전체 입원 기간을 확인하고, ICU 그룹과 비교

In [ ]:
import pandas as pd
import sqlite3
import numpy as np

db_path = '/content/drive/MyDrive/Colab Notebooks/2025-2 데이터애널리틱스/MIMIC4-hosp-icu.db'

# Admissions 테이블 로드
# admittime dischtime deathtime 컬럼 로드
try:
    conn = sqlite3.connect(db_path)
    query = "SELECT hadm_id, admittime, dischtime, deathtime FROM admissions"
    df_admissions = pd.read_sql_query(query, conn)
    conn.close()
    print(f"admissions 테이블 로드 완료 Shape {df_admissions.shape}")
except Exception as e:
    print(f"DB 연결 실패 {e}")

cohort_ver21_hadm = cohort_ver21_hadm.copy()

# is_icu_stay 컬럼 재생성
if 'icu_intime' in cohort_ver21_hadm.columns:
    cohort_ver21_hadm['icu_intime'] = pd.to_datetime(cohort_ver21_hadm['icu_intime'], errors='coerce')
    cohort_ver21_hadm['is_icu_stay'] = cohort_ver21_hadm['icu_intime'].notna().astype(int)
    print("is_icu_stay 컬럼 재생성 완료")
else:
    print("icu_intime 컬럼 부재로 is_icu_stay 생성 불가")

# hadm_id 타입 통일
cohort_ver21_hadm['hadm_id'] = cohort_ver21_hadm['hadm_id'].astype('Int64')
df_admissions['hadm_id'] = df_admissions['hadm_id'].astype('Int64')

# 기존 컬럼 중복 방지를 위한 drop 후 merge
cols_to_drop = ['admittime', 'dischtime', 'deathtime', 'los_days']
cohort_ver21_hadm = cohort_ver21_hadm.drop(columns=[c for c in cols_to_drop if c in cohort_ver21_hadm.columns])

df_merged = pd.merge(cohort_ver21_hadm, df_admissions, on='hadm_id', how='left')

# 시간 컬럼 형식 변환
time_cols = ['admittime', 'dischtime', 'deathtime']
for col in time_cols:
    df_merged[col] = pd.to_datetime(df_merged[col], errors='coerce')

# [1] 병원 재원 일수 los_days 계산
df_merged['los_days'] = (df_merged['dischtime'] - df_merged['admittime']).dt.total_seconds() / (60 * 60 * 24)

print("\n보완 1 los_days 계산 결과 is_icu_stay 별")
if 'is_icu_stay' in df_merged.columns:
    print(df_merged.groupby('is_icu_stay')['los_days'].describe())
else:
    print("is_icu_stay 컬럼 미존재")

# [2] 사망 시점 hours_to_death 재계산
# 우선순위 deathtime > dischtime > dod
if 'dod' in df_merged.columns:
    df_merged['dod'] = pd.to_datetime(df_merged['dod'], errors='coerce')
    df_merged['death_time_final'] = df_merged['deathtime'].fillna(df_merged['dischtime']).fillna(df_merged['dod'])
else:
    df_merged['death_time_final'] = df_merged['deathtime'].fillna(df_merged['dischtime'])

df_merged['hours_to_death'] = (df_merged['death_time_final'] - df_merged['admittime']).dt.total_seconds() / 3600


admissions 테이블 로드 완료 Shape (546028, 4)
is_icu_stay 컬럼 재생성 완료

보완 1 los_days 계산 결과 is_icu_stay 별
              count      mean       std       min       25%       50%  \
is_icu_stay                                                             
0             465.0  4.728079  4.892555 -0.471528  2.304861  3.115972   
1            1465.0  5.679014  9.766460 -0.700694  1.740972  3.165278   

                  75%         max  
is_icu_stay                        
0            4.991667   42.797917  
1            6.508333  219.908333  


### (2) 시술 중 사망하거나 수술실로 직행해서 사망했는지 확인

In [ ]:

# 일반 병동 Non-ICU 사망자 상세 확인
if 'is_icu_stay' in df_merged.columns:
    non_icu_deaths = df_merged[
        (df_merged['is_icu_stay'] == 0) &
        ((df_merged['death_flag'] == 1) | (df_merged['disposition'] == 'EXPIRED'))
    ].copy()

    print(f"\n일반 병동 Non-ICU 사망: {len(non_icu_deaths)}명")

    if not non_icu_deaths.empty:
        # 시술 시도 후 24시간 내 사망자
        cath_deaths = non_icu_deaths[
            (non_icu_deaths['door_to_cath_min'].notna()) &
            (non_icu_deaths['hours_to_death'] < 24)
        ]

        print(f"\nCath 시도 후 24시간 내 사망: {len(cath_deaths)}명")

        if len(cath_deaths) > 0:
            print("\n상세 시술 중 사망 의심 환자")
            print(cath_deaths[['subject_id', 'door_to_cath_min', 'hours_to_death']])
    else:
        print("일반 병동 사망자 없음")


일반 병동 Non-ICU 사망: 102명

Cath 시도 후 24시간 내 사망: 4명

상세 시술 중 사망 의심 환자
     subject_id  door_to_cath_min  hours_to_death
49     10612446             104.0           17.40
252    12858718             173.0            4.70
633    17664919              51.0            2.30
671    18111768             380.0            4.75


### cohort_ver22_complete_STEMI.csv 저장

In [ ]:
import numpy as np
import pandas as pd
import sqlite3

# 1. DB 연결 설정
db_path = '/content/drive/MyDrive/Colab Notebooks/2025-2 데이터애널리틱스/MIMIC4-hosp-icu.db'

# 2. Admissions 테이블에서 병원 입/퇴원 시간 가져오기
try:
    conn = sqlite3.connect(db_path)
    # admittime(입원), dischtime(퇴원), deathtime(사망) 가져오기
    query = "SELECT hadm_id, admittime, dischtime, deathtime FROM admissions"
    df_admissions = pd.read_sql_query(query, conn)
    conn.close()
    print(f"✅ admissions 테이블 로드 완료 (Shape: {df_admissions.shape})")

    # hadm_id 타입 통일
    df_admissions['hadm_id'] = df_admissions['hadm_id'].astype('Int64')

    # 시간 형식 변환
    for col in ['admittime', 'dischtime', 'deathtime']:
        df_admissions[col] = pd.to_datetime(df_admissions[col], errors='coerce')

except Exception as e:
    print(f"⚠️ DB 연결 실패: {e}")

# 3. 데이터 병합 (cohort_ver22 + admissions)
# 방금 만드신 cohort_ver22를 사용합니다.
df_final = cohort_ver22.copy()
df_final['hadm_id'] = df_final['hadm_id'].astype('Int64')

# 혹시 기존에 잘못된 admittime 관련 컬럼이 있다면 제거하고 새로 붙입니다.
cols_to_drop = ['admittime', 'dischtime', 'deathtime', 'los_days']
df_final = df_final.drop(columns=[c for c in cols_to_drop if c in df_final.columns])

# 병합 (Left Join)
df_final = pd.merge(df_final, df_admissions, on='hadm_id', how='left')

# 4. 병원 재원 일수 (los_days) 및 사망 시간 재계산
# (1) 병원 재원 일수 (퇴원 - 입원)
df_final['los_days'] = (df_final['dischtime'] - df_final['admittime']).dt.total_seconds() / (60 * 60 * 24)

# (2) 입원 후 사망까지 시간 (hours_to_death) 업데이트
# admittime이 새로 들어왔으므로 정확도를 위해 업데이트
# death_time_final이 이미 있다면 사용하고, 없으면 재정의
if 'death_time_final' not in df_final.columns:
    if 'dod' in df_final.columns:
        df_final['dod'] = pd.to_datetime(df_final['dod'])
        df_final['death_time_final'] = df_final['deathtime'].fillna(df_final['dischtime']).fillna(df_final['dod'])
    else:
        df_final['death_time_final'] = df_final['deathtime'].fillna(df_final['dischtime'])

df_final['hours_to_death'] = (df_final['death_time_final'] - df_final['admittime']).dt.total_seconds() / 3600

# 5. 컬럼 정리 및 확인
print("\n✅ 최종 병합 및 계산 완료")
print(f"총 컬럼 수: {len(df_final.columns)}")

# 확인: admittime과 los_days가 있는지 확인
check_cols = ['admittime', 'dischtime', 'los_days', 'icu_los_days']
print(f"필수 컬럼 존재 여부: {[c for c in check_cols if c in df_final.columns]}")


✅ admissions 테이블 로드 완료 (Shape: (546028, 4))

✅ 최종 병합 및 계산 완료
총 컬럼 수: 30
필수 컬럼 존재 여부: ['admittime', 'dischtime', 'los_days', 'icu_los_days']


In [ ]:
import pandas as pd

# 1. 작업할 데이터프레임 복사
df_filtered = cohort_ver22.copy()

# 2. 컬럼명 매핑 (현재 데이터의 컬럼명 -> 요청하신 리스트의 이름)
# 코드를 실행하면서 생성된 이름들을 요청하신 이름으로 변경합니다.
rename_map = {
    'icu_los_days': 'icu_los',           # 생성된 재원일수 변수명 변경
    'first_ecg_time': 'first_stemi_ecg_time' # 원본 컬럼명 변경 (필요 시)
}
df_filtered.rename(columns=rename_map, inplace=True)

# 3. 남길 컬럼 리스트 정의 (요청하신 Index 리스트)
target_columns = [
    'subject_id', 'hadm_id', 'icu_stay_id', 'icu_intime', 'icu_outtime',
    'gender', 'race', 'arrival_transport', 'disposition',
    'first_stemi_ecg_time', 'door_to_ecg_min', 'admittime', 'lab_tat_min',
    'door_to_antithrombotic_min', 'door_to_cath_min', 'boarding_delay_min',
    'death_flag', 'icu_los', 'transfer_type', 'is_emergency_admission',
    'dod', 'eventtype', 'cohort_source'
]

# 4. 필터링 수행
# 데이터프레임에 실제로 존재하는 컬럼만 교집합으로 선택합니다.
existing_cols = [col for col in target_columns if col in df_filtered.columns]
missing_cols = [col for col in target_columns if col not in df_filtered.columns]

df_final_selected = df_filtered[existing_cols]

# 5. 결과 확인
print(f"✅ 컬럼 필터링 완료")
print(f"- 선택된 컬럼 수: {len(df_final_selected.columns)}개")
print(f"- 제거된 파생 변수 예시: is_icu_stay, hours_to_death, los_days(병원재원) 등")

if missing_cols:
    print(f"\n⚠️ 주의: 요청 리스트 중 다음 컬럼은 현재 데이터에 없어 포함되지 않았습니다:\n   {missing_cols}")
    print("   (이 컬럼들이 필요하다면 이전 병합 단계에서 누락되었는지 확인해주세요.)")

# 6. 상위 5개 행 출력 및 저장
print("\n[최종 데이터프레임 샘플]")
print(df_final_selected.head())

✅ 컬럼 필터링 완료
- 선택된 컬럼 수: 24개
- 제거된 파생 변수 예시: is_icu_stay, hours_to_death, los_days(병원재원) 등

[최종 데이터프레임 샘플]
   subject_id   hadm_id  icu_stay_id          icu_intime         icu_outtime  \
0    10010058  26359957     39919549 2147-11-18 00:50:00 2147-11-18 03:19:00   
1    10012438  22764825     35750515 2178-06-07 19:33:00 2178-06-07 21:57:00   
2    10013310  21243435     31736720 2153-05-26 08:56:00 2153-05-26 14:18:39   
3    10013310  27682188     35160955 2153-05-06 18:28:00 2153-05-07 20:47:19   
4    10033552  26487381     32542054                 NaT                 NaT   

  gender                      race arrival_transport disposition  \
0      M  ASIAN - SOUTH EAST ASIAN         AMBULANCE    ADMITTED   
1      M                     WHITE         AMBULANCE    ADMITTED   
2      F             BLACK/AFRICAN           UNKNOWN    ADMITTED   
3      F             BLACK/AFRICAN         AMBULANCE    ADMITTED   
4      F                     OTHER         AMBULANCE    ADMITTED   

  fi

In [ ]:
import os

# 1. 저장할 기본 경로 설정
base_path = '/content/drive/MyDrive/Colab Notebooks/2025-2 데이터애널리틱스'

# 2. 파일명 및 전체 경로 결합
file_name = "cohort_ver22_complete_STEMI.csv"
full_save_path = os.path.join(base_path, file_name)

# 3. 파일로 저장 (인코딩: utf-8-sig 권장)
try:
    df_final_selected.to_csv(full_save_path, index=False, encoding='utf-8-sig')
    print(f"\n💾 최종 데이터가 성공적으로 저장되었습니다.")
    print(f"   📂 저장 경로: {full_save_path}")

except OSError as e:
    print(f"\n⚠️ 경로 오류 발생: {e}")
    print("   구글 드라이브가 마운트되어 있는지, 해당 폴더가 존재하는지 확인해주세요.")


💾 최종 데이터가 성공적으로 저장되었습니다.
   📂 저장 경로: /content/drive/MyDrive/Colab Notebooks/2025-2 데이터애널리틱스/cohort_ver22_complete_STEMI.csv
